File: 735-asteroid-collision/asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();
        
        for (int asteroid : asteroids) {
            boolean add = true;
            while (!stack.isEmpty() && asteroid < 0 && stack.peek() > 0) {
                if (stack.peek() < -asteroid) {
                    stack.pop();
                    continue;
                } else if (stack.peek() == -asteroid) {
                    stack.pop();
                }
                add = false;
                break;
            }
            if (add) {
                stack.push(asteroid);
            }
        }
        
        int[] result = new int[stack.size()];
        for (int i = result.length - 1; i >= 0; i--) {
            result[i] = stack.pop();
        }
        return result;
    }
}

File: 735-asteroid-collision/asteroid-collision.c
#include <stdlib.h>
#include <string.h>

int* asteroidCollision(int* asteroids, int asteroidsSize, int* returnSize) {
    int* stack = (int*)malloc(asteroidsSize * sizeof(int));
    int top = -1;
    
    for (int i = 0; i < asteroidsSize; i++) {
        int asteroid = asteroids[i];
        while (top >= 0 && asteroid < 0 && stack[top] > 0) {
            if (stack[top] < -asteroid) {
                top--;
                continue;
            } else if (stack[top] == -asteroid) {
                top--;
            }
            asteroid = 0;
            break;
        }
        if (asteroid != 0) {
            stack[++top] = asteroid;
        }
    }
    
    *returnSize = top + 1;
    return stack;
}